src/System.Linq/src/System/Linq/Set.cs

```diff
@@ -191,6 +191,38 @@ internal List<TElement> ToList()
         }
 
         /// <summary>
+        /// Converts this set to a <see cref="HashSet{T}"/>.
+        /// </summary>
+        /// <param name="comparer">The equality comparer for the new <see cref="HashSet{T}"/>.</param>
+        /// <returns>A <see cref="HashSet{T}"/> of the items in this set.</returns>
+        internal HashSet<TElement> ToHashSet(IEqualityComparer<TElement> comparer)
```

Comment:
I'm inclined toward keeping it. Once the test for the list-provider interface has been made there's no further runtime cost of optimising further, and I think it's clear enough what's happening so the cost to contributor comprehension isn't high either IMO. In for a penny…